File: SlopShopDense/services/payments-csharp/Pagination.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SlopShop.Catalog
{
    public sealed class Page<T>
    {
        private const int DefaultSize = 20;
        private const int MaxSize = 100;

        public IReadOnlyList<T> Items { get; }
        public int Number { get; }
        public int Size { get; }
        public int Total { get; }

        private Page(IReadOnlyList<T> items, int number, int size, int total)
        {
            Items = items;
            Number = number;
            Size = size;
            Total = total;
        }

        public int Pages => Size == 0 ? 0 : (Total + Size - 1) / Size;

        public bool HasNext => Number < Pages;

        public bool HasPrevious => Number > 1;

        public static int NormalizeSize(int requested)
            => requested <= 0 ? DefaultSize : Math.Min(requested, MaxSize);

        public static Page<T> Of(IReadOnlyList<T> all, int number, int size)
        {
            number = Math.Max(1, number);
            size = NormalizeSize(size);
            var window = all.Skip((number - 1) * size).Take(size).ToList();
            return new Page<T>(window, number, size, all.Count);
        }
    }
}

File: SlopShopDense/services/catalog-java/src/main/java/io/slopshop/catalog/Pagination.java
package io.slopshop.catalog;

import java.util.Collections;
import java.util.List;

/** Immutable pagination window. */
public final class Pagination<T> {
    private static final int DEFAULT_SIZE = 20;
    private static final int MAX_SIZE = 100;

    private final List<T> items;
    private final int page;
    private final int size;
    private final int total;

    private Pagination(List<T> items, int page, int size, int total) {
        this.items = items;
        this.page = page;
        this.size = size;
        this.total = total;
    }

    public static int normalizeSize(int requested) {
        if (requested <= 0) return DEFAULT_SIZE;
        return Math.min(requested, MAX_SIZE);
    }

    public static int normalizePage(int requested) {
        return Math.max(1, requested);
    }

    public static <T> Pagination<T> of(List<T> all, int page, int size) {
        page = normalizePage(page);
        size = normalizeSize(size);
        int from = Math.min((page - 1) * size, all.size());
        int to = Math.min(from + size, all.size());
        return new Pagination<>(all.subList(from, to), page, size, all.size());
    }

    public List<T> items() {
        return Collections.unmodifiableList(items);
    }

    public int page() { return page; }

    public int size() { return size; }

    public int total() { return total; }

    public int pages() {
        return size == 0 ? 0 : (total + size - 1) / size;
    }

    public boolean hasNext() {
        return page < pages();
    }
}
